onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu

```diff
@@ -72,29 +57,110 @@ __global__ void SkipLayerNormKernel(
   LayerNorm<T, TPB>(thread_data, ld, offset, beta, gamma, epsilon, output);
 }
 
+// Vectorized kernel
+template <typename T, unsigned TPB, int ILP>
+__global__ void SkipLayerNormKernelSmall(
+    const int ld, const T* input, const T* skip, const T* beta, const T* gamma,
+    const T* bias, const T epsilon, T* output, bool hasBias) {
+  const T rld = T(1.f / ld);
+  const int idx = blockIdx.x * ld + threadIdx.x * ILP;  // grid_size = n / ld
+
+  using VecT = aligned_vector<T, ILP>;
+  __shared__ T mu;      // mean
+  __shared__ T rsigma;  // 1 / std.dev.
+
+  T input_v[ILP], skip_v[ILP], bias_v[ILP], output_v[ILP];
+
+  VecT* input_val = reinterpret_cast<VecT*>(&input_v);
+  *input_val = *reinterpret_cast<const VecT*>(&input[idx]);
+
+  VecT* skip_val = reinterpret_cast<VecT*>(&skip_v);
+  *skip_val = *reinterpret_cast<const VecT*>(&skip[idx]);
+
+  if (hasBias) {
+    VecT* bias_val = reinterpret_cast<VecT*>(&bias_v);
+    *bias_val = *reinterpret_cast<const VecT*>(&bias[threadIdx.x * ILP]);
+  }
+
+  cub::KeyValuePair<T, T> thread_data(T(0.f), T(0.f));
+
+  if (ILP * threadIdx.x < ld) {
+    T rldval_sum = T(0.f);
+    T rldvalsq_sum = T(0.f);
+    #pragma unroll
+    for (int i = 0; i < ILP; i++) {
+      input_v[i] += hasBias ? skip_v[i] + bias_v[i]: skip_v[i];
+      const T rldval = rld * input_v[i];
+      rldval_sum += rldval;
+      rldvalsq_sum += rldval * input_v[i];
+    }
+    thread_data = cub::KeyValuePair<T, T>(rldval_sum, rldvalsq_sum);
+  }
+  LayerNormSmall<T, TPB, ILP>(input_v, thread_data, ld, idx, beta, gamma, epsilon, output);
+}
+
+/* half16 and float32 */
 template <typename T>
 bool ComputeSkipLayerNorm(
-    cudaStream_t stream, const int ld, const int n, const T* input, const T* skip,
-    const T* beta, const T* gamma, const T* bias, const T epsilon, T* output) {
+    cudaStream_t stream, const int ld, const int n, const T* input, const T* skip, const T* beta,
+    const T* gamma, const T* bias, const T epsilon, T* output) {
   // this must be true because n is the total size of the tensor
   assert(n % ld == 0);
-  const int grid_size = n / ld;
-
-  if (ld <= 32) {
-    constexpr int block_size = 32;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
-  } else if (ld <= 128) {
-    constexpr int block_size = 128;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
-  } else if (ld == 384) {
-    constexpr int block_size = 384;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
+  bool hasBias = (bias == nullptr) ? false : true;
+  if (0 == (n % 4)) {
+    const int grid_size = n / ld;
+    if (ld <= 32) {
```

Comment:
A block size smaller than warp size (32 for CUDA, 64 for AMD) probably doesn't make sense. We need to think about this more.